atc/db/pipeline_lifecycle.go

```diff
@@ -0,0 +1,89 @@
+package db
+
+import (
+	sq "github.com/Masterminds/squirrel"
+	"github.com/concourse/concourse/atc/db/lock"
+)
+
+//go:generate counterfeiter . PipelineLifecycle
+
+type PipelineLifecycle interface {
+	ArchiveAbandonedPipelines() (int, error)
+}
+
+func NewPipelineLifecycle(conn Conn, lockFactory lock.LockFactory) PipelineLifecycle {
+	return &pipelineLifecycle{
+		conn:        conn,
+		lockFactory: lockFactory,
+	}
+}
+
+type pipelineLifecycle struct {
+	conn        Conn
+	lockFactory lock.LockFactory
+}
+
+func (pl *pipelineLifecycle) ArchiveAbandonedPipelines() (int, error) {
+	tx, err := pl.conn.Begin()
+	if err != nil {
+		return 0, err
+	}
+
+	defer Rollback(tx)
+
+	latestSuccessfulBuilds := `WITH successful_builds AS (
+      SELECT DISTINCT ON (job_id) id, job_id
+      FROM builds
+      WHERE status='succeeded'
+      ORDER BY job_id, id DESC
+      )`
+	rows, err := pipelinesQuery.
+		Prefix(latestSuccessfulBuilds).
+		LeftJoin("jobs j ON j.id = p.parent_job_id").
+		LeftJoin("successful_builds ON successful_builds.job_id = p.parent_job_id").
+		Where(sq.Or{
+			// pipeline is no longer set by the parent pipeline
+			sq.And{
+				sq.NotEq{"parent_build_id": nil},
```

Comment:
I'm also wondering if we can add an index to this column? I don't see one added in the migration that introduced the column but maybe there's a reason for not adding one? Also same with the `archived` boolean.

https://github.com/concourse/concourse/blob/master/atc/db/migration/migrations/1591737494_add_parent_job_id.up.sql

Reply:
We added an index to the `parent_build_id` and `parent_job_id`. Not sure if they're really needed though because we did an `EXPLAIN ANALYZE` and postgres never used either of the indexes. Not sure if it was due to the small dataset we had or not. wdyt?

Reply:
If there are no rows that are fetched using the index or if the dataset was small enough that postgres determined it was faster to not use an idex, postgres explain analyze won't show that it uses the index. I just ran the query for archiving pipelines after a build is finished (https://github.com/concourse/concourse/pull/5854/files#diff-cd2dc5a827d8e3673211bb2ac1540801R556-R567) and it seemed to be using the index for `parent_job_id` but not `parent_build_id`

```
  CTE pipelines_to_archive
    ->  Recursive Union  (cost=4.16..178.36 rows=81 width=4) (actual time=0.055..0.967 rows=3 loops=1)
          ->  Bitmap Heap Scan on pipelines  (cost=4.16..9.51 rows=1 width=4) (actual time=0.036..0.063 rows=1 loops=1)
                Recheck Cond: (parent_job_id = 1)
                Filter: ((NOT archived) AND (parent_build_id < 2))
                Heap Blocks: exact=1
                ->  Bitmap Index Scan on parent_job_id  (cost=0.00..4.16 rows=2 width=0) (actual time=0.015..0.023 rows=1 loops=1)
                      Index Cond: (parent_job_id = 1)
          ->  Hash Join  (cost=1.42..16.72 rows=8 width=4) (actual time=0.233..0.274 rows=1 loops=3)
                Hash Cond: (j.pipeline_id = pipelines_to_archive.id)
                ->  Hash Join  (cost=1.09..16.29 rows=8 width=8) (actual time=0.098..0.174 rows=3 loops=3)
                      Hash Cond: (p_1.parent_job_id = j.id)
                      ->  Seq Scan on pipelines p_1  (cost=0.00..14.10 rows=410 width=8) (actual time=0.010..0.050 rows=4 loops=3)
                      ->  Hash  (cost=1.04..1.04 rows=4 width=8) (actual time=0.115..0.123 rows=4 loops=1)
                            Buckets: 1024  Batches: 1  Memory Usage: 9kB
                            ->  Seq Scan on jobs j  (cost=0.00..1.04 rows=4 width=8) (actual time=0.019..0.061 rows=4 loops=1)
                ->  Hash  (cost=0.20..0.20 rows=10 width=4) (actual time=0.042..0.045 rows=1 loops=3)
                      Buckets: 1024  Batches: 1  Memory Usage: 9kB
                      ->  WorkTable Scan on pipelines_to_archive  (cost=0.00..0.20 rows=10 width=4) (actual time=0.008..0.019 rows=1 loops=3)
  ->  Seq Scan on pipelines p  (cost=0.00..14.10 rows=410 width=167) (actual time=0.018..0.057 rows=4 loops=1)
  ->  Hash  (cost=1.62..1.62 rows=81 width=4) (actual time=1.072..1.079 rows=3 loops=1)
        Buckets: 1024  Batches: 1  Memory Usage: 9kB
        ->  CTE Scan on pipelines_to_archive pa  (cost=0.00..1.62 rows=81 width=4) (actual time=0.072..1.037 rows=3 loops=1)
Planning Time: 0.339 ms
Execution Time: 1.531 ms
```

I'm not exactly sure why it isn't using the index for `parent_build_id`, maybe it's due to the fact that its not an equal operation but is actually a less than operation? Or it could be due to my dataset being too small that it was cheaper to not use the index. 

In any case, we can keep the indexes and monitor the usage in `hush-house`. If we find that the indexes are not useful, we can remove them.

Reply:
I'm guessing just a `parent_job_id` index enough. I'm guessing it'll be rare that a single job would have enough child pipelines to fill more than a few database pages, which is likely why Postgres chooses not to use the `parent_build_id` index.

Reply:
Ah yeah thats a good point! I definitely think we can remove the `parent_build_id` index then.